README.md
# ![[logo](https://github.com/gaojiuli/xweb)](logo.png)


![[Build](https://travis-ci.org/gaojiuli/xweb)](https://travis-ci.org/gaojiuli/xweb.svg?branch=master)
![[License](https://pypi.python.org/pypi/xweb/)](https://img.shields.io/pypi/l/xweb.svg)
![[Pypi](https://pypi.python.org/pypi/xweb/)](https://img.shields.io/pypi/v/xweb.svg)
![[Python](https://pypi.python.org/pypi/xweb/)](https://img.shields.io/pypi/pyversions/xweb.svg)

High performance web framework built with uvloop and httptools

In Xweb, everything is asynchronous.

## Features

-  High performance.
-  Asynchronous.
-  Small.

## Requirements

- Python3.6+

## Installation

`pip install xweb`

## Get Started

### Hello World

```python
from xweb import App

app = App()

@app.use
async def response(ctx):
    ctx.res.body = "Hello World"


if __name__ == '__main__':
    app.listen(8000)
```

### Example with middleware.

A middleware is an async function or an async callable object which looks like: `async def logger(ctx, fn)`

```python
# app.py

import time

from xweb import App

app = App()


@app.use
async def logger(ctx, fn):
    await fn()
    rt = ctx['X-Response-Time']
    print(rt)


@app.use
async def response_time(ctx, fn):
    start = time.time()
    await fn()
    usage = (time.time() - start) * 1000_000
    ctx['X-Response-Time'] = f'{usage:.0f}µs'


@app.use
async def response(ctx):
    ctx.res.body = "Hello World"


if __name__ == '__main__':
    app.listen(8000)
```

## App

- app.use(fn)
- app.listen(host='127.0.0.1', port=8000, debug=True)

## Context

- ctx.req
- ctx.res
- ctx.send
- ctx.abort(self, status, msg="", properties="")
- ctx.check(self, value, status=400, msg='', properties="")

### Request
    
`ctx.req` is a Request object.

- ctx.req.headers dict
- ctx.req.method str
- ctx.req.url str
- ctx.req.raw bytes
- ctx.req.ip str

### Response

`ctx.res` is a Request object.

- ctx.res.body str
- ctx.res.status int
- ctx.res.msg str
- ctx.res.headers dict



## Benchmark

- Benchmark code in benchmarks/.
- environment: `iMac (Retina 4K, 21.5-inch, 2017)`, `3 GHz Intel Core i5`, `8 GB 2400 MHz DDR4`
- test command: `wrk http://127.0.0.1:8000/ -c 100 -t 10 -d 10 -T 10`

Frameworks| Requests/Sec | Version
-----|-----|-----
xweb|100000|0.1.1
vibora|90000|0.0.6
meinheld + wsgi|77000|0.6.1
sanic|50000|0.7.0

## Deploy and Run

`python app.py`.

## Test

1. `pip install -r requirement.txt`
2. `pytest --cov xweb.py`

## Contributing


### Build Middleware.

XWeb is inspired by [koajs](https://koajs.com/). I need some help for writing middleware as in koa. For example:

1. Body parser. Convert the raw bytes body into dict or file.
2. Data validator. Async data validator with high performance.
3. Router. High performance router like koa-router.
4. etc..

### Open issue.

1. Suggestion.
2. Bug.

CHANGELOG.md
## 2018.09.29

1. Init

xweb.py
import asyncio
import logging
import multiprocessing
import os
import socket
from email.utils import formatdate
from functools import partial
from http import HTTPStatus
from inspect import signature

import httptools

try:
    import uvloop

    asyncio.set_event_loop_policy(uvloop.EventLoopPolicy())

except ImportError:
    pass

FORMAT = '[%(asctime)-15s] %(message)s'
logging.basicConfig(format=FORMAT)
logger = logging.getLogger('xweb')
logger.setLevel(logging.DEBUG)

__version__ = '0.1.3'
__author__ = 'Jiuli Gao'
__all__ = ('Request', 'Response', 'App', 'HTTPException', 'Context', 'HTTPStatus', 'logger')


class HTTPException(Exception):
    def __init__(self, status, msg, properties):
        self.properties = properties
        self.msg = msg
        self.status = status


class Request:
    def __init__(self):
        self.headers = {}
        self.method = "HEAD"
        self.url = "/"
        self.raw = None
        self.ip = None

    def __getitem__(self, item):
        return getattr(self, item)

    def __setitem__(self, key, value):
        setattr(self, key, value)


class Response:
    def __init__(self):
        self.body = "Hello Xweb!"
        self.status = 200
        self.msg = ""
        self.headers = {
            'Date': formatdate(timeval=None, localtime=False, usegmt=True),
            'Content-Type': 'text/plain'
        }

    def __getitem__(self, item):
        return getattr(self, item)

    def __setitem__(self, key, value):
        setattr(self, key, value)

    def __bytes__(self):
        http_status = HTTPStatus(self.status)
        http_status_raw = f"HTTP/1.1 {http_status.value} {http_status.phrase}\r\n".encode()
        http_body_raw = self.body.encode()
        self.headers['Content-Length'] = len(http_body_raw)
        http_header_raw = "".join([f'{k}: {v}\r\n' for k, v in self.headers.items()]).encode() + b'\r\n'
        return http_status_raw + http_header_raw + http_body_raw


class Context:
    def __init__(self):
        self.req = Request()
        self.res = Response()
        self.write = None

    def send(self, _):
        self.write(bytes(self.res))

    def check(self, value, status=400, msg='', properties=""):
        if not value:
            self.abort(status=status, msg=msg, properties=properties)

    def abort(self, status, msg="", properties=""):
        raise HTTPException(status=status, msg=msg, properties=properties)

    def __getattr__(self, item):
        return getattr(self.req, item)

    def __getitem__(self, item):
        return getattr(self, item)

    def __setitem__(self, key, value):
        setattr(self, key, value)

    @property
    def body(self):
        return self.res.body

    @body.setter
    def body(self, value):
        self.res.body = value

    @property
    def status(self):
        return self.res.status

    @status.setter
    def status(self, value):
        self.res.status = value

    @property
    def msg(self):
        return self.res.msg

    @msg.setter
    def msg(self, value):
        self.res.msg = value


class HTTPProtocol(asyncio.Protocol):

    def __init__(self, handler, loop):
        self.parser = None
        self.transport = None
        self.handler = handler
        self.loop = loop
        self.ctx = None

    def connection_made(self, transport):
        self.parser = httptools.HttpRequestParser(self)
        self.transport = transport

    def on_url(self, url):
        self.ctx = Context()
        client = self.transport.get_extra_info('peername')
        self.ctx.req.ip = client[0]
        self.ctx.write = self.transport.write
        self.ctx.req.url = url.decode()

    def on_header(self, name, value):
        self.ctx.req.headers[name.decode()] = value.decode()

    def on_body(self, body):
        self.ctx.req.raw += body

    def on_message_complete(self):
        task = self.loop.create_task(self.handler(self.ctx))
        task.add_done_callback(self.ctx.send)

    def data_received(self, data):
        self.parser.feed_data(data)

    def connection_lost(self, exc):
        self.transport.close()


class App:
    def __init__(self):
        self.handlers = []
        self.workers = set()

    def use(self, fn):
        self.handlers.append(fn)

    def serve(self, sock):
        loop = asyncio.new_event_loop()
        server = loop.create_server(partial(HTTPProtocol, loop=loop, handler=self), sock=sock)
        loop.create_task(server)
        try:
            loop.run_forever()
        except KeyboardInterrupt:
            server.close()
            loop.close()

    def listen(self, port=8000, host="127.0.0.1", workers=multiprocessing.cpu_count()):
        pid = os.getpid()
        print(f'[{pid}] Starting xweb {__version__}')
        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
        sock.setblocking(False)
        sock.bind((host, port))
        os.set_inheritable(sock.fileno(), True)

        try:
            print(f'[{pid}] Listening at: http://{host}:{port}')
            print(f'[{pid}] Workers: {workers}')
            for _ in range(workers):
                worker = multiprocessing.Process(target=self.serve, kwargs=dict(sock=sock))
                worker.daemon = True
                worker.start()
                print(f'[{pid}] Starting worker with pid: {worker.pid}')
                self.workers.add(worker)
            for worker in self.workers:
                worker.join()
        except KeyboardInterrupt:
            print('\r', end='\r')
            print(f'[{pid}] Server soft stopping')
            for worker in self.workers:
                worker.terminate()
                worker.join()
            print(f'[{pid}] Server stopped successfully!')
        sock.close()

    async def __call__(self, ctx):
        if not self.handlers:
            ctx.abort(404)
        next_fn = None

        for handler in self.handlers[::-1]:
            if len(signature(handler).parameters) == 2:
                next_fn = partial(handler, ctx=ctx, fn=next_fn)
            else:
                next_fn = partial(handler, ctx=ctx)
        try:
            await next_fn()
        except HTTPException as e:
            ctx.res.status = e.status
            ctx.res.body = e.msg or HTTPStatus(e.status).phrase
            ctx.res.msg = e.properties
